app/services/katello/applicability/applicable_content_helper.rb

```diff
@@ -0,0 +1,91 @@
+module Katello
+  module Applicability
+    class ApplicableContentHelper
+      attr_accessor :content_facet, :content_unit_class
+
+      def initialize(content_unit_class, content_facet = nil)
+        self.content_facet = content_facet
+        self.content_unit_class = content_unit_class
+      end
+
+      def calculate_and_import
+        if self.content_facet.bound_repositories.any?
+          to_add, to_remove = applicable_differences
+          ActiveRecord::Base.transaction do
+            insert(to_add) unless to_add.blank?
+            remove(to_remove) unless to_remove.blank?
+          end
+        end
+      end
+
+      def content_ids
+        bound_library_instance_repos = self.content_facet.bound_repositories.collect do |repo|
+          repo.library_instance.nil? ? repo : repo.library_instance
+        end
+
+        if self.content_unit_class == ::Katello::Erratum
+          # Query for all Errata ids that are attached to the host's applicable packages
+          return ::Katello::Erratum.joins("INNER JOIN katello_repository_errata ON \
+          katello_errata.id = katello_repository_errata.erratum_id").
+          where(id: ::Katello::ErratumPackage.where(nvrea: self.content_facet.applicable_rpms.select(:nvra)).
+                select(:erratum_id)).where("katello_repository_errata.repository_id in \
+                (#{bound_library_instance_repos.pluck(:id).join(',')})").pluck(:id).uniq
+        elsif self.content_unit_class == ::Katello::ModuleStream
```

Comment:
This errata query seems fairly slow for me, this was much faster:

```
          query = 'select katello_repository_errata.erratum_id  as id from katello_repository_errata
                       inner join katello_erratum_packages
                          on  katello_repository_errata.erratum_id  = katello_erratum_packages.erratum_id
                       inner join katello_rpms
                          on katello_rpms.nvra = katello_erratum_packages.nvrea
                       inner join katello_content_facet_applicable_rpms on
                           katello_content_facet_applicable_rpms.rpm_id = katello_rpms.id
                       where
                            "katello_content_facet_applicable_rpms"."content_facet_id" = :content_facet_id
                       AND katello_repository_errata.repository_id IN (:repo_ids)'

          Katello::Erratum.find_by_sql([query, {content_facet_id: content_facet.id, repo_ids: bound_library_instance_repos.pluck(:id)} ]).map(&:id).uniq

```

Creating an index on katello_rpms (nvra) seemed to help with the previous query, but backing it seems to not have a noticable effect with the new query, something to keep an eye on